### PR TITLE
Add Limit to Concurrent Promises

### PR DESCRIPTION
**Pull request by Dylan700**
I've attempted to fix the issues here (#481 #490) by limiting the number of concurrent promises that can occur at a given time, using the [p-limit](https://www.npmjs.com/package/p-limit) package.

I've come accross these issues myself where it seems files are being skipped and not uploaded due to resource limits being reached with too many Node listeners being added etc. Hopefully this will help fix the problem.

I've also added the option for the user to specify the concurrency limit in the `Client()` constructor, however, a better approach may be to include this in the connection method options. I'll leave this decision up to the maintainer.

### CODE REVIEW

**Comment by theophilusx** [comment]
Thanks for the PR. I am travelling at the moment and without reliable internet access. However, I will look at this when I return in October.Sent from my iPadOn 27 Jul 2023, at 7:17 am, Dylan700 ***@***.***> wrote:﻿I've attempted to fix the issues here (#481 #490) by limiting the number of concurrent promises that can occur at a given time, using the p-limit package.
I've come accross these issues myself where it seems files are being skipped and not uploaded due to resource limits being reached with too many Node listeners being added etc. Hopefully this will help fix the problem.
I've also added the option for the user to specify the concurrency limit in the Client() constructor, however, a better approach may be to include this in the connection method options. I'll leave this decision up to the maintainer.



You can view, comment on, or merge this pull request online at:
  https://github.com/theophilusx/ssh2-sftp-client/pull/492

Commit Summary

  06b4e1f  Add Limit to Concurrent Promises


File Changes  (2 files)

  
    M
    package.json
    (1)
  
  
    M
    src/index.js
    (12)
  


Patch Links:

  https://github.com/theophilusx/ssh2-sftp-client/pull/492.patch
  https://github.com/theophilusx/ssh2-sftp-client/pull/492.diff

**Comment by Dylan700** [comment]
Hi @theophilusx, just wondering if you've had a chance to look at this yet?

**Comment by Dylan700** [comment]
@theophilusx, any updates on this?

**Comment by theophilusx** [comment]
with some luck, I'll be looking at this over the next week.

**Comment by theophilusx** [comment]
Have now had an initial look at this patch. The first issue which jumps out at me is that the added dependency on the p-limit package could be an issue. The latest version of that package v5 is ESM only. You cannot use CJS require to import it. This means, either we have to migrate to ESM (not desirable at this time) or we have to use a kludge and dynamic import of p-limit or we revert to version 3.x or we find another alternative which does not rely on p-limit. . As other bugs have been fixed in v4 and v5, I'm not keen on reverting to v3.  Using v5 also means a minimal node support verison of v18 and p-limit v5 require node v18.  I was planning on updating the minimum to v16

So, basically, I have to either do the questionable kludge to handle a dynamic async import of p-limit or find another alternative which does not rely on p-limit. The first option would be quicker, but the second one is likely more maintainable.  I also note there are a couple of issues relating to problems with ESM moduels for deno and node, which is another reaosn I'm not keen on switching to ESM now, even though it is inevitable for the long term. As ssh2, the main package this module relies on is still CJS, I'd rather stick with CJS for now.

**Comment by Dylan700** [comment]
Thanks for looking into this. I've found this alternative library: https://www.npmjs.com/package/es6-promise-pool
It seems to be inactive (the last commit was 3 years ago) but it allows CJS.

Let me know your thoughts on this.

**Comment by theophilusx** [comment]
Thinking about the issue, I think we can solve it very simply without
the need for any additional library. All we really need to do is put all
the source and destination arguments into an array, the partition the
array by the max concurrency value and then iterate through the
partitioned array, executing each 'block' of src/dest with a
Promise.all, collect the results in an array and return it as the result
for the encompassing control promise which wraps the whole thing. This
will limit the maximum concurrent promises and still provide some
performance benefits.

This will only need a slight refactoring of the donwloadDir/uploadDir
functions and will avoid adding an additional dependency to the
module. I'll work on that and adding some better tests for
downloadDir/uploadDir which will push the concurrency limits
etc. Should be able to get this done this week given the year is
beginning to slow down with the end of year break approaching.

**Comment by Dylan700** [comment]
That sounds like a much more maintainable and simpler solution! I'll close up this PR.
